codec/src/main/java/io/netty/handler/codec/BulkLengthFieldBasedFrameDecoder.java

```diff
@@ -0,0 +1,160 @@
+        }
+
+        if (lengthFieldOffset > maxFrameLength - lengthFieldLength) {
+            throw new IllegalArgumentException(
+                    "maxFrameLength (" + maxFrameLength + ") " +
+                    "must be equal to or greater than " +
+                    "lengthFieldOffset (" + lengthFieldOffset + ") + " +
+                    "lengthFieldLength (" + lengthFieldLength + ").");
+        }
+
+        this.maxFrameLength = maxFrameLength;
+        this.lengthFieldOffset = lengthFieldOffset;
+        this.lengthFieldLength = lengthFieldLength;
+        lengthFieldEndOffset = lengthFieldOffset + lengthFieldLength;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) {
+        int readableBytes = buffer.readableBytes();
+        int rdIdx = buffer.readerIndex();
+
+        while (readableBytes >= lengthFieldEndOffset) {
+            int actualLengthFieldOffset = rdIdx + lengthFieldOffset;
+            long frameLength = getUnadjustedFrameLength(buffer, actualLengthFieldOffset, lengthFieldLength);
+
+            if (frameLength < 0) {
+                throw new CorruptedFrameException(
+                        "negative pre-adjustment length field: " + frameLength);
+            }
+
+            frameLength += lengthFieldEndOffset;
+
+            if (frameLength > maxFrameLength) {
+                throw new TooLongFrameException(
+                        "Adjusted frame length exceeds " + maxFrameLength +
+                        ": " + frameLength + " - discarded");
+            }
+
+            int frameLengthInt = (int) frameLength;
+
+            if (readableBytes < frameLengthInt) {
+                break;
+            }
+
+            readableBytes -= frameLengthInt;
+            rdIdx += frameLengthInt;
+        }
+
+        if (rdIdx == 0) {
+            return;
+        }
+
+        final int previousWriterIndex = buffer.writerIndex();
+
+        buffer.writerIndex(rdIdx);
+
+        final ByteBuf chunk = ctx.alloc().buffer(rdIdx, rdIdx).writeBytes(buffer, rdIdx);
+
+        buffer.writerIndex(previousWriterIndex);
+        buffer.discardReadBytes();
+
+        out.add(chunk);
+    }
+
+    protected long getUnadjustedFrameLength(ByteBuf buf, int offset, int length) {
+        long frameLength;
+        switch (length) {
+        case 1:
+            frameLength = buf.getUnsignedByte(offset);
+            break;
+        case 2:
+            frameLength = buf.getUnsignedShort(offset);
+            break;
+        case 3:
+            frameLength = buf.getUnsignedMedium(offset);
+            break;
+        case 4:
+            frameLength = buf.getUnsignedInt(offset);
+            break;
+        case 8:
+            frameLength = buf.getLong(offset);
+            break;
+        default:
+            throw new DecoderException(
+                    "unsupported lengthFieldLength: " + lengthFieldLength + " (expected: 1, 2, 3, 4, or 8)");
```

Comment:
I don't think so. Either way you have to call an appropriate method which depends on an offset value. The same approach is used in LengthFieldBasedFrameDecoder. Seems legit for me.